### ISSUE TITLE
regression tests

### ISSUE THREAD

**Issue by liruipeng** [hypre-space/hypre#290]
Running regression tests on GPUs is almost done with a few exceptions that I am not sure how to deal with

1. SStruct `neighborpart.sh`. `.out.30, 31, 32` are supposed to be exactly the same. But on GPUs, `.30, .31` and `.32` are not, with error
```
[32] (2) - Final Relative Residual Norm = 4.226312e-07
[30,31] (2) + Final Relative Residual Norm = 4.498034e-07      (err 2.72e-08)
```
2. SStruct `periodic.sh`. `.out.80, 81,.. 85` are supposed to be exactly the same. But on GPUs, `80` and `82, 85` are not

```
(2) - Final Relative Residual Norm = 1.301852e-16
(2) + Final Relative Residual Norm = 7.850462e-17      (err 5.17e-17)

(2) - Final Relative Residual Norm = 1.301852e-16
(2) + Final Relative Residual Norm = 7.850462e-17      (err 5.17e-17)
```
Also, `90` and `92, 95` are not
```
(2) - Final Relative Residual Norm = 2.923987e-17
(2) + Final Relative Residual Norm = 3.123203e-17      (err 1.99e-18)

(2) - Final Relative Residual Norm = 2.923987e-17
(2) + Final Relative Residual Norm = 3.123203e-17      (err 1.99e-18)
```

Since it seems doing some sanity check, so they ought to be identical maybe, instead of roundings. I am not sure if they are actually real problems.

Any suggestions?

**Comment by rfalgout** [comment]
Hi @liruipeng .  For `neighborpart.sh`, I think the numbers should match to bitwise precision.  This is a 2-part problem with the parts stacked on top of each other and the matrix storage is SSTRUCT, so A = As + Au, where As is structured and Au is ParCSR.  Since this uses a finite element assembly interface and the coefficients are constant, then the matrices should be bitwise the same in all cases (the row/col numbers of the nonzeroes in Au may differ, but they should be ordered the same relative to each other, so I think the computations should occur in the same order in all cases as well).  Also, the ordering for the ParCSR part is the same in all cases, which means that the matvec should produce bitwise identicial results in all cases (StructMatvec always does this).  The solver 8 is a split preconditioner with Jacobi applied to the structured part, so the whole thing should be bitwise identical.

For the `periodic.sh` test, I think the only question would be the inner product in the CG iteration and possibly the matvec in solver 28 since it uses a ParCSR matrix.  The other solvers basically use Struct for the matvec.  The final relative residuals are pretty small here, which is a bit strange since this is a 2D Laplacian.  As a sanity check, maybe you should try running the same problem with the Struct driver and use the `-print` option to verify the matrix/rhs values in both cases.

Hope this helps.

**Comment by liruipeng** [comment]
> Hi @liruipeng . For `neighborpart.sh`, I think the numbers should match to bitwise precision. This is a 2-part problem with the parts stacked on top of each other and the matrix storage is SSTRUCT, so A = As + Au, where As is structured and Au is ParCSR. Since this uses a finite element assembly interface and the coefficients are constant, then the matrices should be bitwise the same in all cases (the row/col numbers of the nonzeroes in Au may differ, but they should be ordered the same relative to each other, so I think the computations should occur in the same order in all cases as well). Also, the ordering for the ParCSR part is the same in all cases, which means that the matvec should produce bitwise identicial results in all cases (StructMatvec always does this). The solver 8 is a split preconditioner with Jacobi applied to the structured part, so the whole thing should be bitwise identical.
> 
> For the `periodic.sh` test, I think the only question would be the inner product in the CG iteration and possibly the matvec in solver 28 since it uses a ParCSR matrix. The other solvers basically use Struct for the matvec. The final relative residuals are pretty small here, which is a bit strange since this is a 2D Laplacian. As a sanity check, maybe you should try running the same problem with the Struct driver and use the `-print` option to verify the matrix/rhs values in both cases.
> 
> Hope this helps.

@rfalgout Thank you for the suggestions. Correction:   For `neighborpart.sh`, it was `.32` that differs from `30, 31`, which uses `4` MPI ranks, whereas the others use `2`. If np is different, I assume in Au, `diag` and `offd` are different, how column indices are split and how they are ordered. But I am not sure since you said "the ordering for the ParCSR part is the same in all cases". Since the matvecs are done in parallel, across rows and within a row, is it possible they don't give the same result? Thanks!

**Comment by liruipeng** [comment]
@rfalgout Also, for `periodic.sh`, since `-solver 28` uses csr-matvec while `218/18` uses struct matvec, we should not expect they give bit-wise consistent results. Is it correct?

**Comment by rfalgout** [comment]
HI @liruipeng .  I think the 4-proc run should also give the same results.  In the 2-proc runs, each part lives on a process.  In run 31, each part is split in two in the x-direction, which changes the global ordering of the unknowns, but maintains the same ordering along the part boundaries.  In the 4-proc run 32, each part is again split in two in the x-direction, so the global ordering of unknowns is the same as for run 31 (unknowns are ordered first by process, then by block, then lexicographically within each block).  The only difference is that there is extra parallelism.

The Au matrix does not have a diagonal, just off-diagonal coefficients that couple the two parts together.  I'm pretty sure those coefficients should be bitwise the same in all cases, and since the ordering along the interface is also the same in run 32 as in the other cases, the Au matvec should also be very accurate if not bitwise equal.  The only thing that might change the order of computations in the Au matvec for 32 are diagonal stencil couplings right where the four procs intersect (in the middle of the domain) maybe due to the way communicated data is processed.  I don't know the ParCSR matvec code well enough to know how that is implemented, but I don't see why the computations should be different anywhere else.

This test was designed to check the equivalence of these runs, so we need to be sure we understand why it fails before making any changes to the test.

**Comment by rfalgout** [comment]
> @rfalgout Also, for `periodic.sh`, since `-solver 28` uses csr-matvec while `218/18` uses struct matvec, we should not expect they give bit-wise consistent results. Is it correct?

Correct.  What I don't understand here is the tiny final relative residual, because this is basically a 4-minus-1 Laplacian problem with rhs=1.  It looks like I added these tests.  I'm not sure why the relative residual of zero in the saved file didn't bother me at the time.  I'll look into that a bit more.

**Comment by rfalgout** [comment]
Hi @liruipeng .  Interestingly, the periodic tests 80-95 all converge in exact arithmetic in 4 iterations for 80-85 and 5 iterations for 90-95.  I don't have a complete explanation yet, but I'll let you know when I do.  It's clear that exact convergence should happen in no more than 8 and 9 iterations respectively, since the problem is periodic in y, has only 8 points in x for 80-85 and 9 points in x for 90-95, and the rhs is one with initial guess zero.  As a result, understanding convergence reduces to understanding convergence of CG for a 1D Lapacian on a grid with 8 or 9 points, because each grid line in x solves the same problem with the same initial guess and the same iterates.

Would you try running 80-85 manually but without the `-rhsone` comand-line option?  This will use a different right-hand-side and should produce the following in all cases:

    Iterations = 46
    Final Relative Residual Norm = 7.750679e-07

For 90-95 you should get the following in all cases:

    Iterations = 48
    Final Relative Residual Norm = 8.594333e-07

The previous tests probably only work when things are bitwise accurate, which isn't guaranteed here because of ParCSR and also the use of MPI for the inner products.  We may have more luck with this on GPUs and in general.  Let me know.

I am thinking about changing these anyway.  I also noticed I left out runs 84 and 94 from the `.out` file by accident.

**Comment by rfalgout** [comment]
Hi @liruipeng .  Here is the explanation for exact convergence of CG in 4 or 5 iterations.  Consider a 1D Laplacian on an 8-point grid with right-hand side 1 and initial guess 0.  The Krylov space after 4 iterations is full rank over half of the domain: the residual r is all ones, Ar is nonzero right next to the boundary (and zero elsewhere), A^2r is nonzero at two points next to the boundary, and A^3r is nonzero at 3 points next to the boundary.  The basis over the other half of the domain is symmetric to this, but since the initial error is symmetric, the Krylov space can be used to fully correct the error over the entire domain.  Because the 2D problem is periodic in y, the solution is constant in the y direction, so by having the same initial guess and rhs for all of the x-line problems, convergence is determined by the 1D case.  The same argument can be used to argue 5 iterations for the 9-pt nodal problem.

Let me know if the changes I suggested help with the tests.  I don't remember if I knew about the exact convergence when I created these tests.  I was probably more focused on looking at the matrices themselves, and maybe added the `-rhsone` option at the last minute to the `.jobs` file and just didn't notice the zero residuals.  Thanks!

**Comment by liruipeng** [comment]
> Hi @liruipeng . Interestingly, the periodic tests 80-95 all converge in exact arithmetic in 4 iterations for 80-85 and 5 iterations for 90-95. I don't have a complete explanation yet, but I'll let you know when I do. It's clear that exact convergence should happen in no more than 8 and 9 iterations respectively, since the problem is periodic in y, has only 8 points in x for 80-85 and 9 points in x for 90-95, and the rhs is one with initial guess zero. As a result, understanding convergence reduces to understanding convergence of CG for a 1D Lapacian on a grid with 8 or 9 points, because each grid line in x solves the same problem with the same initial guess and the same iterates.
> 
> Would you try running 80-85 manually but without the `-rhsone` comand-line option? This will use a different right-hand-side and should produce the following in all cases:
> 
> ```
> Iterations = 46
> Final Relative Residual Norm = 7.750679e-07
> ```
> 
> For 90-95 you should get the following in all cases:
> 
> ```
> Iterations = 48
> Final Relative Residual Norm = 8.594333e-07
> ```
> 
> The previous tests probably only work when things are bitwise accurate, which isn't guaranteed here because of ParCSR and also the use of MPI for the inner products. We may have more luck with this on GPUs and in general. Let me know.
> 
> I am thinking about changing these anyway. I also noticed I left out runs 84 and 94 from the `.out` file by accident.

Hi, @rfalgout 

I ran the tests without `-rhsone`. I got the same results as you showed above. Also, thank you for the explanation for exact convergence with constant rhs vectors.

**Comment by rfalgout** [comment]
I just created a new PR with the modified periodic test.

**Comment by liruipeng** [comment]
Hi, @rfalgout, Finally, I find a possible reason that `neighborpart` job 32 gives different results from those of 30 and 31. In `HYPRE_SStructSplitSolve`, `hypre_SStructInnerProd` is called and thus `struct_innerprod.c`, which generates different results with different `np` on GPUs (because of parallel reduction). Does it make sense? Thanks!

**Comment by rfalgout** [comment]
Hi @liruipeng .  I'm not so sure about this.  If that were the case, I would expect all of them to be different from each other, because the local GPU data has different orderings in 30/31 and it is smaller in size in 32.  It's a good thought and still possible, but the best way to check is to print out the solution data in all three cases and see if they are the same.  The inner products here are only used to compute the residual norm for the convergence test.

Did you look into the diagonal-stencil-for-Au idea I mentioned (for test 32)?  Actually, printing out the results would also help uncover if that is an issue.  Thanks for looking into this!

**Comment by liruipeng** [comment]
> Hi @liruipeng . I'm not so sure about this. If that were the case, I would expect all of them to be different from each other, because the local GPU data has different orderings in 30/31 and it is smaller in size in 32. It's a good thought and still possible, but the best way to check is to print out the solution data in all three cases and see if they are the same. The inner products here are only used to compute the residual norm for the convergence test.
> 
> Did you look into the diagonal-stencil-for-Au idea I mentioned (for test 32)? Actually, printing out the results would also help uncover if that is an issue. Thanks for looking into this!

Hi, @rfalgout here are so far what I have found. 

1. `Au` only has only 1 nonzero per row. So running matvec with it on GPUs and CPUs should give exactly the same result. I have verified by running csr-matvec back on CPU. No changes in the final residual.

2. Running Boxloops from the structured side on CPUs does remove the differences. I speculate only the reduction loops matter. But I haven't further fine tuned it yet. (It is a new feature I just added. But the results make sense)

3. Changing the execution policy of the whole unstructured side doesn't seem to matter. This includes IJ assembly if any. This run doesn't seem to have AMG setup. So, maybe it doesn't change anything by doing this.

I have printed out the solutions, which are split into several files. I actually don't know which one to look at in particular. Should I just `diff` all of them.

Thanks again

-Ruipeng

**Comment by rfalgout** [comment]
Hi @liruipeng .  It's good to know there are no diagonal stencil coefficients.  Regarding the solution printouts, the file entries should start with grid point numbers `p: (grid point; 0)` where `p` is the processor number.  Since the grid points are global within each part, you can take all of the files for a part, cat them together, and use `sort` to get a set of files that can be diffed.  The only real problem is the processor number at the beginning, but I can use emacs to quickly remove them.

Maybe the easiest thing to do is to just tar up the three sets of files and send them to me to look at.  Thanks!

You mention AMG setup.  All of these runs basically just do Jacobi.

**Comment by rfalgout** [comment]
Hi @liruipeng .  Thanks for the files.  The solution for job 32 is different.  I'll try to better understand where it's different and get back to you.

**Comment by rfalgout** [comment]
Hello again @liruipeng .  It looks like there are some differences in the struct part of the matrix on part 0 for run 32, so something must be wrong in the code.  I will do a couple of CPU runs to verify that the problem doesn't exist there (it probably doesn't, since the residuals all match) and let you know what I find.

**Comment by liruipeng** [comment]
This problem should have been fixed in #294 .
